src/main/scala/org/broadinstitute/dsde/firecloud/core/AgoraPermissionHandler.scala

```diff
@@ -0,0 +1,121 @@
+  def toAgoraPermission(fireCloudPermission: FireCloudPermission):AgoraPermission = {
+    AgoraPermission(Some(fireCloudPermission.user), Some(toAgoraRoles(fireCloudPermission.role)))
+  }
+
+  // convenience method to translate an AgoraPermission to a FireCloudPermission object
+  def toFireCloudPermission(agoraPermission: AgoraPermission):FireCloudPermission = {
+    // if the Agora permission has a None/empty/whitespace user, the following will throw
+    // an IllegalArgumentException trying to create the FireCloudPermission. That exception
+    // will be caught elsewhere.
+    FireCloudPermission(agoraPermission.user.getOrElse(""), toFireCloudRole(agoraPermission.roles))
+  }
+
+  // translation between a FireCloud role and a list of Agora roles
+  def toAgoraRoles(fireCloudRole:String) = {
+    fireCloudRole match {
+      case NoAccess => ListNoAccess
+      case Reader => ListReader
+      case Owner => ListOwner // Could use "All" instead but this is more precise
+      case _ => ListNoAccess
+    }
+  }
+
+  // translation between a list of Agora roles and a FireCloud role
+  def toFireCloudRole(agoraRoles:Option[List[String]]) = {
+    agoraRoles match {
+      case None => NoAccess
+      case Some(r) => {
+        r.sorted match {
+          case ListNoAccess => NoAccess
+          case ListReader => Reader
+          case ListOwner => Owner
+          case ListAll => Owner
+          case _ => NoAccess
+        }
+      }
+    }
+  }
+}
+
+class AgoraPermissionActor (requestContext: RequestContext) extends Actor with FireCloudRequestBuilding {
+
+  implicit val system = context.system
+  import system.dispatcher
+
+  val log = Logging(system, getClass)
+
+  val pipeline = authHeaders(requestContext) ~> sendReceive
+
+  def receive = {
+    // GET requests
+    case AgoraPermissionHandler.Get(url: String) =>
+      val permissionListFuture: Future[HttpResponse] = pipeline { Get(url) }
+      handleAgoraResponse(permissionListFuture)
+
+    case AgoraPermissionHandler.Post(url: String, agoraPermissions: List[AgoraPermission]) =>
+      val permissionListFuture: Future[HttpResponse] = pipeline {Post(url, agoraPermissions)}
+      handleAgoraResponse(permissionListFuture)
+
+    case _ =>
+      context.parent ! RequestComplete(StatusCodes.BadRequest)
+      context stop self
+  }
+
+  def handleAgoraResponse(permissionListFuture: Future[HttpResponse]) = {
+    permissionListFuture onComplete {
+      case Success(response) =>
+        response.status match {
+          case StatusCodes.OK =>
+            try {
+              val agoraPermissions = unmarshal[List[AgoraPermission]].apply(response)
+              val fireCloudPermissions = agoraPermissions.map(x => x.toFireCloudPermission)
+              context.parent ! RequestComplete(OK, fireCloudPermissions)
+              context stop self
+            } catch {
+              // TODO: more specific and graceful error-handling
```

Comment:
I'm thinking about this too. Not in scope for this PR, but definitely a tech-debt thing that we should try to address,